packages/next/build/webpack/plugins/next-trace-entrypoints-plugin.ts

```diff
@@ -0,0 +1,253 @@
+              } catch (e) {
+                if (e.code === 'ENOENT' || e.code === 'EISDIR') {
+                  return null
+                }
+                throw e
+              }
+            }
+            const readlink = (path: string) => {
+              try {
+                return compilation.inputFileSystem.readlinkSync(path)
+              } catch (e) {
+                if (
+                  e.code !== 'EINVAL' &&
+                  e.code !== 'ENOENT' &&
+                  e.code !== 'UNKNOWN'
+                ) {
+                  throw e
+                }
+                return null
+              }
+            }
+            const stat = (path: string) => {
+              try {
+                return compilation.inputFileSystem.statSync(path)
+              } catch (e) {
+                if (e.code === 'ENOENT') {
+                  return null
+                }
+                throw e
+              }
+            }
+
+            const nftCache = {}
+            const entryPaths = Array.from(entryModMap.keys())
+
+            for (const entry of entryPaths) {
+              depModMap.clear()
+              const entryMod = entryModMap.get(entry)
+              // TODO: investigate caching, will require ensuring no traced
+              // files in the cache have changed, we could potentially hash
+              // all traced files and only leverage the cache if the hashes
+              // match
+              // const cachedTraces = entryMod.buildInfo?.cachedNextEntryTrace
+
+              // Use cached trace if available and trace version matches
+              // if (
+              //   isWebpack5 &&
+              //   cachedTraces &&
+              //   cachedTraces.version === TRACE_OUTPUT_VERSION
+              // ) {
+              //   this.entryTraces.set(
+              //     entryNameMap.get(entry)!,
+              //     cachedTraces.tracedDeps
+              //   )
+              //   continue
+              // }
+              const collectDependencies = (mod: any) => {
+                if (!mod || !mod.dependencies) return
+
+                for (const dep of mod.dependencies) {
+                  const depMod = getModuleFromDependency(compilation, dep)
+
+                  if (depMod?.resource && !depModMap.get(depMod.resource)) {
+                    depModMap.set(depMod.resource, depMod)
+                    collectDependencies(depMod)
+                  }
+                }
+              }
+              collectDependencies(entryMod)
+
+              const toTrace: string[] = [entry, ...depModMap.keys()]
+
+              const root = nodePath.parse(process.cwd()).root
+              const result = await nodeFileTrace(toTrace, {
+                base: root,
+                cache: nftCache,
+                processCwd: this.appDir,
+                readFile,
+                readlink,
+                stat,
+                ignore: TRACE_IGNORES,
```

Comment:
Do we also need to ignore the user's input from `unstable_excludeFiles` here?

Reply:
I was currently applying this at the end of the build where `unstable_includeFiles` is applied in case an `includeFile` collided with an `excludeFile`

Reply:
I suspect it will impact performance because you can stop tracing earlier so we don't look up deps of deps.

https://github.com/vercel/vercel/blob/18bec983aefbe2a77bd14eda6fca59ff7e956d8b/packages/node/src/index.ts#L204

Reply:
Added this to the ignore config here https://github.com/vercel/next.js/pull/25538/commits/9823923a39a715fc50b4a4ca4955b9b099bd6ceb

Reply:
Or actually it seems we can't include this in the plugin since we need the built pages to gather the page configs 🤔 we might need to make this a `next.config.js` config

Reply:
Oh I assumed thats what it was: a `next.config.js` config so users could exclude files